stumpy/mmotifs.py

```diff
@@ -0,0 +1,186 @@
+        finding the dimension of each motif
+    """
+    T = core._preprocess(T)
+
+    m = T.shape[-1] - P.shape[-1] + 1
+
+    if max_motifs < 1:  # pragma: no cover
+        logger.warning(
+            "The maximum number of motifs, `max_motifs`, "
+            "must be greater than or equal to 1"
+        )
+        logger.warning("`max_motifs` has been set to `1`")
+        max_motifs = 1
+
+    T, M_T, Σ_T = core.preprocess(T, m)
+
+    excl_zone = int(np.ceil(m / config.STUMPY_EXCL_ZONE_DENOM))
+
+    if max_matches is None:
+        max_matches = np.inf
+
+    if cutoffs is None:
+        P_copy = P.copy().astype(np.float64)
+        P_copy[np.isinf(P_copy)] = np.nan
+        cutoffs = np.nanmax(
+            [
+                np.nanmean(P_copy - 2.0 * np.nanstd(P_copy, axis=0), axis=1),
+                np.nanmin(P_copy, axis=1),
+            ],
+            axis=0,
+        )
+    elif isinstance(cutoffs, int) or isinstance(cutoffs, float):
+        cutoffs = np.full(T.shape[0], cutoffs)
+
+    motif_distances = []
+    motif_indices = []
+    motif_subspaces = []
+    motif_mdls = []
+
+    candidate_idx = np.argmin(P, axis=1)
+    nn_idx = I[np.arange(len(candidate_idx)), candidate_idx]
+
+    while len(motif_distances) < max_motifs:
+        mdls, subspaces = mdl(T, m, candidate_idx, nn_idx)
+        k = np.argmin(mdls)
+        motif_mdls.append(mdls)
+
+        motif_idx = candidate_idx[k]
+        motif_value = P[k, motif_idx]
+
+        if (
+            motif_value > cutoffs[k]
+            or not np.isfinite(motif_value)
+            or (isinstance(max_distance, float) and motif_value > max_distance)
+        ):
+            break
+
+        motif_subspaces.append(subspaces[k])
+
+        # if isinstance(max_distance, float) and motif_value > max_distance:
+        #     break
+
+        query_matches = match(
+            Q=T[subspaces[k], motif_idx : motif_idx + m],
+            T=T[subspaces[k]],
+            M_T=M_T[subspaces[k]],
+            Σ_T=Σ_T[subspaces[k]],
+            max_matches=max_matches,
+            max_distance=max_distance,
+            atol=atol,
+            # normalize=normalize,
+            # p=p
+        )
+
+        if len(query_matches) > min_neighbors:
+            motif_distances.append(query_matches[:, 0])
+            motif_indices.append(query_matches[:, 1])
+
+        for idx in query_matches[:, 1]:
+            core.apply_exclusion_zone(P, idx, excl_zone, np.inf)
+        candidate_idx = np.argmin(P, axis=1)
+        nn_idx = I[np.arange(len(candidate_idx)), candidate_idx]
+
+    motif_distances = core._jagged_list_to_array(
+        motif_distances, fill_value=np.nan, dtype=np.float64
+    )
+    motif_indices = core._jagged_list_to_array(
+        motif_indices, fill_value=-1, dtype=np.int64
+    )
+
+    return (
+        motif_distances,
+        motif_indices,
+        np.array(motif_subspaces, dtype=object),
+        np.array(motif_mdls, dtype=object),

```

Comment:
Generally speaking, I prefer to return back the list of items in order of "most important/useful" to "least important/useful". In your opinion, is `motif_mdls` more or less important than `motif_subspaces`? I only ask because I recall that you had wanted to plot the `mdls` and so I figured that it may be relevant.

Reply:
I think this is wonderful! 

Btw, in case it matters, I wasn't trying to suggest that we remove `motif_subspaces` or `motif_mdls` but merely to return them in the order that they will most likely be used. I think you are the perfect user in this case as you've thought deeply about the workflow and I trust you. Are you able to comment on how efficient/obvious/intuitive the current implementation is? Again, our goal in STUMPY isn't to "be everything, for everyone" but to make the hard parts a lot simpler or more intuitive. Do you feel like we've achieved that goal? Is there anything glaring that you'd want to change or discuss within this context?

Reply:
I was kind of overwhelmed by the sheer number of parameters available at the beginning (for the one dimensional case as well) but I don't know how we could make this a bit easier on new users.

Maybe it's just the barrier of entry for new users, so to speak. 😄

Or maybe a way around this could be a sectioned tutorial, one with a focus on the essentials and an advanced one, where the parameters are explained with use cases one at a time.

Reply:
This is super valuable feedback! And I think your suggestion to add a section on what the parameters mean would be a welcomed addition. In fact, technically, I think what's missing is a tutorial on motif discovery using `motifs()`. We briefly mention `stumpy.match` in the [Fast Pattern Matching Tutorial](https://stumpy.readthedocs.io/en/latest/Tutorial_Pattern_Matching.html) but we still have an open issue for [1D Motif Discovery](https://github.com/TDAmeritrade/stumpy/issues/359).

Reply:
Oh and another point I noticed is that I was a bit confused at the beginning that the `motifs()` function not only contains motifs, but also nearest neighbours (i.e. matches). That's the reason why I called the `motif_distances` array `motif_matches_distances`. That was easier for me to understand what exactly the function is doing. 
After company meeting where I explained what I am currenty working on, I noticed that people struggled to differentiate between motifs and their nearest neighbours. I had to explain that a motif consists of TWO separate subsequences and can be found using the matrix profile. And that all other subsequences are matches / neighbours of the motif, which can NOT be found using the matrix profile, but with the distance profile only. I also had the same difficulty in understanding until corresponding with you in #494. 
Therefore I would maybe make it a bit clearer that the user can find motifs AND neighbours with `motifs()` or `mmotifs()`. Or in other words that `motifs()` is doing the same as `matches()` with the extra of finding the motifs.

Reply:
> We briefly mention stumpy.match in the [Fast Pattern Matching Tutorial](https://stumpy.readthedocs.io/en/latest/Tutorial_Pattern_Matching.html) but we still have an open issue for https://github.com/TDAmeritrade/stumpy/issues/359. Perhaps, we should/could combine the 1D motif discovery tutorial with the multi-dimensional one.

I don't know how easy it would be to find a suitable dataset that fits both. I would start with the multi-dimensional one since it is relevant for my thesis and see if there is time left for the one-dimensional case while working on it. After graduation I'll certainly have enough time to finish or start the 1D case as well since I'll probably get bored pretty quickly without anything to work on in the summer break 😄

Reply:
> Therefore I would maybe make it a bit clearer that the user can find motifs AND neighbours with motifs() or mmotifs(). Or in other words that motifs() is doing the same as matches() with the extra of finding the motifs.

I get your point now! Do you think it is sufficient to explain this in the `Return` descriptions in the docstring? Or should we go back to `motif_matches_distance` and `motif_matches_indices`? If anything, this just needs to be changed int `motifs.py`, `aamp_motifs.py`, and (later) `aamp_mmotifs.py`

Reply:
> I don't know how easy it would be to find a suitable dataset that fits both. I would start with the multi-dimensional one since it is relevant for my thesis and see if there is time left for the one-dimensional case while working on it. After graduation I'll certainly have enough time to finish or start the 1D case as well since I'll probably get bored pretty quickly without anything to work on in the summer break 😄

Absolutely no pressure. This is volunteer work so I am at your mercy and there is no rush. I think we wouldn't rely on a single dataset and, instead, have a 1-D one as well as a multi-dimensional dataset. Let's keep it in mind for the future.

Reply:
> Do you think it is sufficient to explain this in the Return descriptions in the docstring? Or should we go back to motif_matches_distance and motif_matches_indices? 

In my view it was a bit clearer but I think that writing it in the return description should also suffice.